docs/design/generics/details.md

```diff
@@ -4323,9 +4323,13 @@ class Optional(T:! type) {
 ```
 
 This prevents any higher-priority impl that overlaps a final impl from being
-defined. Further, if the Carbon compiler sees a matching final impl, it can
-assume it won't be specialized so it can use the assignments of the associated
-types in that impl definition.
+defined unless it agrees with the `final` impl on the overlap. Since we do not
```

Comment:
Is this well-defined? Given two arbitrary `impl` declarations, can we always compute what the overlap is, and whether the associated constants have the same values in the overlapping region? (I think the answer might be yes if all parameters are generic and no if there are any template parameters, but I'm not sure.)

What do we want to do for overlap between a templated impl and a final impl / between a templated final impl and a generic impl? That case seems problematic both because there can be template constraints that make the overlap non-computable, and the values of associated constants might depend on the template parameters, which perhaps might make them not symbolically comparable. Do we delay the diagnostic until the template is instantiated?

Reply:
I believe the overlap between generic and concrete `impl` declarations can be computed using unification. I've added a sentence about templated impls being delayed.